zcash_client_sqlite/src/lib.rs

```diff
@@ -457,21 +538,67 @@ impl<'a, P: consensus::Parameters> WalletWrite for DataConnStmtCache<'a, P> {
         })
     }
 
-    fn store_received_tx(
+    fn store_decrypted_tx(
         &mut self,
-        received_tx: &ReceivedTransaction,
+        d_tx: &DecryptedTransaction,
+        nullifiers: &[(AccountId, Nullifier)],
     ) -> Result<Self::TxRef, Self::Error> {
         self.transactionally(|up| {
-            let tx_ref = wallet::put_tx_data(up, received_tx.tx, None)?;
+            let tx_ref = wallet::put_tx_data(up, d_tx.tx, None)?;
 
-            for output in received_tx.outputs {
+            let mut spending_account_id: Option<AccountId> = None;
+            for output in d_tx.sapling_outputs {
                 if output.outgoing {
-                    wallet::put_sent_note(up, output, tx_ref)?;
+                    wallet::put_sent_note(
+                        up,
+                        tx_ref,
+                        output.index,
+                        output.account,
+                        &output.to,
+                        Amount::from_u64(output.note.value)
+                            .map_err(|_| SqliteClientError::CorruptedData("Note value invalid.".to_string()))?,
+                        Some(&output.memo),
+                    )?;
                 } else {
+                    match spending_account_id {
+                        Some(id) =>
+                            if id != output.account {
+                                panic!("Unable to determine a unique account identifier for z->t spend.");
+                            }
+                        None => {
+                            spending_account_id = Some(output.account);
+                        }
+                    }
+
                     wallet::put_received_note(up, output, tx_ref)?;
                 }
             }
 
+            // If we have some transparent outputs:
+            if !d_tx.tx.transparent_bundle().iter().any(|b| b.vout.is_empty()) {
+                // If there are no shielded spends, then this is t2t and we can safely ignore it
+                // otherwise, this is z2t and it might have originated from our wallet.
+                // Store received z->t transactions in the same way they would be stored by
+                // create_spend_to_address. If there are any of our shielded inputs, we interpret
+                // this as our z->t tx and store the vouts as our sent notes.
+                // FIXME this is a weird heuristic that is bound to trip us up somewhere.
```

Comment:
I think that what this is attempting to do is detect shielded spends targeting external transparent recipients and restore those as sent UTXOs when recovering from seed? It doesn't have anything to do with auto-shielding, it looks like.

Reply:
I think that the "safely ignore it" bit of the comment is a bit misleading, and was part of what confused me.

Reply:
In a pairing, @nuttycom, @daira and I concurred that this was fine to keep in (to handle edge cases in the scanning logic for detecting z->t transactions created outside of this wallet), and that this will also likely be rewritten when the new scanning logic is implemented.